pigosat.go

```diff
@@ -233,6 +234,178 @@ func (p *Pigosat) AddClauses(clauses Formula) {
+		return false
+	}
+	// int picosat_failed_assumption (PicoSAT *, int lit);
+	return C.picosat_failed_assumption(p.p, C.int(lit)) != 0
+}
+
+// Returns a list of failed assumption in the last call to Solve(). It only
+// makes sense if the last call to Solve() returned Unsatisfiable.
+func (p *Pigosat) FailedAssumptions() []Literal {
+	defer p.ready(true)()
+	if p.Res() != Unsatisfiable {
+		return nil
+	}
+
+	// const int * picosat_failed_assumptions (PicoSAT *);
+	litPtr := C.picosat_failed_assumptions(p.p)
+	return p.litArrayToSlice(litPtr)
+}
+
+// Converts a 0-terminated array of literal results to a slice.
+// Does not acquire internal locks.
+func (p *Pigosat) litArrayToSlice(litPtr *C.int) []Literal {
+	if litPtr == nil || *litPtr == 0 {
+		return []Literal{}
+	}
+	// It should be reasonable to use the number of vars in
+	// the solver as the max array size, since we aren't tracking
+	// the active number of assumptions.
+	size := int(C.picosat_variables(p.p))
+
+	var cints []C.int
+	header := (*reflect.SliceHeader)(unsafe.Pointer(&cints))
+	header.Cap = size
+	header.Len = size
+	header.Data = uintptr(unsafe.Pointer(litPtr))
+
+	// The returned int pointer is both temporary, and larger than
+	// needed, so we need to copy the real values into a new slice,
+	// up until the terminator.
+	ints := []Literal{}
+	for _, cint := range cints {
+		// break at the first sign of the 0 terminator.
+		if cint == 0 {
+			break
+		}
+		ints = append(ints, Literal(cint))
+	}
+	return ints
+}
+
+// Compute one maximal subset of satisfiable assumptions. You need to set
+// the assumptions, call 'Solve()' before calling this function.
+// The result is a list of assumptions that consistently can be asserted
+// at the same time.  Before returing the library 'reassumes' all assumptions.
+//
+// It could be beneficial to set the default phase of assumptions
+// to true (positive).  This can speed up the computation.
+func (p *Pigosat) MaxSatisfiableAssumptions() []Literal {
+	defer p.ready(false)()
+	if C.picosat_inconsistent(p.p) != 0 {
+		return []Literal{}
+	}
+	//const int * picosat_maximal_satisfiable_subset_of_assumptions (PicoSAT *);
+	litPtr := C.picosat_maximal_satisfiable_subset_of_assumptions(p.p)
+	return p.litArrayToSlice(litPtr)
+}
+
+// This function assumes that you have set up all assumptions with
+// 'Assume()'.  Then it calls 'Solve()' internally unless the
+// formula is already inconsistent without assumptions, i.e.  it contains
+// the empty clause.  After that it extracts a maximal satisfiable subset of
+// assumptions.
+//
+// The result is a zero terminated maximal subset of consistent assumptions
+// or an empty list if the formula contains the empty clause and thus no
+// more maximal consistent subsets of assumptions can be extracted.  In the
+// first case, before returning, a blocking clause is added, that rules out
+// the result for the next call.
+//
+// NOTE: adding the blocking clause changes the CNF.
```

Comment:
TBH I didn't look to much into the `blocksol` stuff too closely. I see now that its a Pigosat addition that implicitly blocks on Solves. But if I understand this correctly, this shouldn't impact what we have done in our go `Solve()` method, right? This will call into C, which will use the C solve and then block the solution. That would mean the only custom logic that has been layered in Pigosat is the blocking being done from our own Solve